File: update_note_dates.py
#!/usr/bin/env python3
"""pre-commitフック(.githooks/pre-commit)から呼ばれ、ステージ済みの
notes/src/*.md にYAML frontmatterで作成日・更新日を刻印する。

- frontmatterが無い新規ノート → created/updated を追加
- frontmatterがある既存ノート → updated だけ現在時刻に更新 (createdは保持)

ファイルのmtimeには依存しない (clone直後やCIではmtimeが壊れるため)。
"""

import datetime
import re
import sys
from pathlib import Path

now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")

for name in sys.argv[1:]:
    path = Path(name)
    if not path.is_file():
        continue
    content = path.read_text(encoding="utf-8")
    m = re.match(r"\A---\n(.*?)\n---\n", content, re.S)
    if m:
        fm = m.group(1)
        if re.search(r"^updated:\s*.+$", fm, re.M):
            fm = re.sub(r"^updated:\s*.+$", "updated: %s" % now, fm, flags=re.M)
        else:
            fm += "\nupdated: %s" % now
        content = "---\n%s\n---\n%s" % (fm, content[m.end():])
    else:
        content = "---\ncreated: %s\nupdated: %s\n---\n%s" % (now, now, content)
    path.write_text(content, encoding="utf-8")
    print("Stamped %s" % path)

File: scripts/update-note-dates.mjs
// pre-commitフック (.githooks/pre-commit) から呼ばれ、ステージ済みの
// notes/src/*.md に YAML frontmatter で作成日・更新日を刻印する。
//
// - frontmatter が無い新規ノート → created/updated を追加
// - frontmatter がある既存ノート → updated だけ現在時刻に更新 (created は保持)
//
// ファイルの mtime には依存しない (clone直後やCIでは mtime が壊れるため)。
import { existsSync, readFileSync, writeFileSync } from 'node:fs';

const p = (n) => String(n).padStart(2, '0');
const d = new Date();
const now = `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}`;

for (const file of process.argv.slice(2)) {
  if (!existsSync(file)) continue;
  let content = readFileSync(file, 'utf8');
  const m = content.match(/^---\n([\s\S]*?)\n---\n/);
  if (m) {
    let fm = m[1];
    fm = /^updated:\s*.+$/m.test(fm)
      ? fm.replace(/^updated:\s*.+$/m, `updated: ${now}`)
      : `${fm}\nupdated: ${now}`;
    content = `---\n${fm}\n---\n${content.slice(m[0].length)}`;
  } else {
    content = `---\ncreated: ${now}\nupdated: ${now}\n---\n${content}`;
  }
  writeFileSync(file, content, 'utf8');
  console.log(`Stamped ${file}`);
}
